### PR TITLE
FeatureTable difference_lag

### PR DESCRIPTION
**Pull request by yaxinchen666**

### CODE REVIEW

File: pyzoo/zoo/friesian/feature/utils.py [hunk 1]
```diff
@@ -96,3 +96,26 @@ def ordinal_shuffle_partition(df):
...
+    elif isinstance(columns, list):
+        return name_sep.join(columns)
+    else:
+        raise ValueError("item should be either str or list of str")
+
+
+def str_to_list(arg, arg_name):
+    if isinstance(arg, str):
+        return [arg]
+    assert isinstance(arg, list), arg_name + " should be str or list"
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/utils.py:120]: str or a list of str

File: pyzoo/zoo/friesian/feature/table.py [hunk 2]
```diff
@@ -1150,6 +1151,62 @@ def split(self, ratio, seed=None):
...
         return tuple(tbl_list)
 
+    def difference_lag(self, columns, sort_cols, shifts=1, partition_cols=None, out_cols=None):
+        """
+        Calculates the difference between two consecutive rows, or two rows with certain interval
+        of the specified continuous columns. The table is first partitioned by partition_cols if it
+        is not None, and then sorted by sort_cols before the calculation.
+        :param columns: str or list of str. Continuous columns to calculate the difference.
+        :param sort_cols: str or list of str. Columns by which the table is sorted.
+        :param shifts: int or list of int. Intervals between two rows.
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:1161]: a list of

File: pyzoo/zoo/friesian/feature/table.py [hunk 3]
```diff
@@ -1150,6 +1151,62 @@ def split(self, ratio, seed=None):
...
+        assert not nonnumeric_col_type, \
+            "columns should be numeric but get " + \
+            ", ".join(list(map(lambda x: x[0] + " of type " + x[1], nonnumeric_col_type)))
+        if isinstance(shifts, int):
+            shifts = [shifts]
+        elif isinstance(shifts, list):
+            for s in shifts:
+                assert isinstance(s, int), "elements in shift should be integer but get " + str(s)
+        else:
+            raise ValueError("shift should be either int or list of int")
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:1180]: a list of int

File: pyzoo/zoo/friesian/feature/table.py [hunk 4]
```diff
@@ -1150,6 +1151,62 @@ def split(self, ratio, seed=None):
...
+        if partition_cols is not None:
+            partition_cols = str_to_list(partition_cols, "partition_cols")
+        if out_cols is None:
+            sort_name = gen_cols_name(sort_cols)
+            out_cols = [[sort_name + "_dl_" + column + "_" + str(shift)
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:1185]: dl -> diff_lag?

File: pyzoo/zoo/friesian/feature/table.py [hunk 5]
```diff
@@ -1150,6 +1151,62 @@ def split(self, ratio, seed=None):
...
+                         for shift in shifts] for column in columns]
+        else:
+            assert isinstance(out_cols, list), "out_cols should be list of list of str"
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:1188]: why out_cols is a list of list?

File: pyzoo/zoo/friesian/feature/table.py [hunk 6]
```diff
@@ -1158,10 +1158,10 @@ def target_encode(self, cat_cols, target_cols, target_mean=None, smooth=20, kfol
         For each categorical column / column group in cat_cols, calculate the mean of target
         columns in target_cols and encode the Table with mean.
 
-        :param cat_cols: str, or list of (str or list of str). Categorical columns / column groups
-               to target encode. If an element in the list is a str, then it is a categorical
+        :param cat_cols: str, or a list of (str or list of str). Categorical columns / column
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:1161]: str, a list of str or a nested list of str.

File: pyzoo/zoo/friesian/feature/table.py [hunk 7]
```diff
@@ -1177,9 +1177,9 @@ def target_encode(self, cat_cols, target_cols, target_mean=None, smooth=20, kfol
                range [0, kfold). Default is "__fold__".
         :param drop_cat: Boolean. Drop the categorical columns if it is true. Default is True.
         :param drop_folds: Boolean. Drop the fold column if it is true. Default is True.
-        :param out_cols: list of list of str. Each inner list corresponds to the categorical column
-               in the same position of cat_cols. Each element in the inner list corresponds to the
-               target column in the same position of target_cols. If it is None, the output
+        :param out_cols: a list of list of str. Each inner list corresponds to the categorical
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:1180]: a nested list of str

File: pyzoo/zoo/friesian/feature/table.py [hunk 8]
```diff
@@ -1322,11 +1322,11 @@ def encode_target(self, targets, target_cols=None, drop_cat=True):
         """
         Encode columns with provided list of TargetCode.
 
-        :param cat_cols: str, or list of (str or list of str). Categorical columns / column groups
-               to target encode. If an element in the list is a str, then it is a categorical
+        :param cat_cols: str, or a list of (str or list of str). Categorical columns / column
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:1325]: same as above

File: pyzoo/zoo/friesian/feature/table.py [hunk 9]
```diff
@@ -1355,6 +1355,83 @@ def encode_target(self, targets, target_cols=None, drop_cat=True):
 
         return result_tbl
 
+    def difference_lag(self, columns, sort_cols, shifts=1, partition_cols=None, out_cols=None):
+        """
+        Calculates the difference between two consecutive rows, or two rows with certain interval
+        of the specified continuous columns. The table is first partitioned by partition_cols if it
+        is not None, and then sorted by sort_cols before the calculation.
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:1362]: empty line

File: pyzoo/zoo/friesian/feature/table.py [hunk 10]
```diff
@@ -1355,6 +1355,83 @@ def encode_target(self, targets, target_cols=None, drop_cat=True):
...
+        :param columns: str or a list of str. Continuous columns to calculate the difference.
+        :param sort_cols: str or a list of str. Columns by which the table is sorted.
+        :param shifts: int or a list of int. Intervals between two rows.
+        :param partition_cols: Columns by which the table is partitioned.
+        :param out_cols: str, a list of str, or a list of list of str. When both columns and shifts
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:1367]: nested

File: pyzoo/zoo/friesian/feature/table.py [hunk 11]
```diff
@@ -1355,6 +1355,83 @@ def encode_target(self, targets, target_cols=None, drop_cat=True):
...
+               has only one element, out_cols can be a str. When columns or shifts has only one
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:1368]: a str -> a single str

File: pyzoo/zoo/friesian/feature/table.py [hunk 12]
```diff
@@ -1355,6 +1355,83 @@ def encode_target(self, targets, target_cols=None, drop_cat=True):
...
+               element, out_cols can be a list of str, and each element in out_cols corresponds to
+               an element in shifts or columns. When it is a list of list of str, each inner list
+               corresponds to a column in columns. Each element in the inner list corresponds to a
+               shift in shifts. If it is None, the output column will be sort_cols + "_diff_lag_"
+               + column + "_" + shift. Default is None.
+        :return: a new Table with difference columns.
```
Comment (hkvision) [inline pyzoo/zoo/friesian/feature/table.py:1374]: A new FeatureTable ...

**Comment by yaxinchen666** [comment]
Jenkins passed:
http://10.112.231.51:18888/job/ZOO-PR-Validation/6166/
http://10.112.231.51:18888/job/ZOO-PR-Python-Spark-2.3-py37/3007/
http://10.112.231.51:18888/job/ZOO-PR-Validation-Other/6230/
http://10.112.231.51:18888/job/ZOO-PR-Python-ExampleTests-py37-Spark3/1703/

**Pull request merged**